consensus/src/pending_order_votes.rs

```diff
@@ -0,0 +1,176 @@
+// Copyright © Aptos Foundation
+// Parts of the project are originally copyright © Meta Platforms, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use aptos_consensus_types::{common::Author, order_vote::OrderVote};
+use aptos_crypto::{hash::CryptoHash, HashValue};
+use aptos_logger::prelude::*;
+use aptos_types::{
+    aggregate_signature::PartialSignatures,
+    ledger_info::{LedgerInfo, LedgerInfoWithPartialSignatures, LedgerInfoWithSignatures},
+    validator_verifier::{ValidatorVerifier, VerifyError},
+};
+use std::{collections::HashMap, sync::Arc};
+
+/// Result of the order vote processing. The failure case (Verification error) is returned
+/// as the Error part of the result.
+#[derive(Debug, PartialEq, Eq)]
+pub enum OrderVoteReceptionResult {
+    /// The vote has been added but QC has not been formed yet. Return the amount of voting power
+    /// QC currently has.
+    VoteAdded(u128),
+    /// The very same vote message has been processed in past.
+    DuplicateVote,
+    /// The very same author has already voted for another proposal in this round (equivocation).
+    EquivocateVote,
+    /// This block has just been certified after adding the vote.
+    NewLedgerInfoWithSignatures(Arc<LedgerInfoWithSignatures>),
+    /// There might be some issues adding a vote
+    ErrorAddingVote(VerifyError),
+    /// The vote is not for one of the last 2 rounds
+    UnexpectedRound(u64, u64),
+    /// Error happens when aggregating signature
+    ErrorAggregatingSignature(VerifyError),
+}
+
+#[derive(Debug, PartialEq, Eq)]
+enum OrderVoteStatus {
+    EnoughVotes,
+    NotEnoughVotes,
+}
+
+/// A PendingVotes structure keep track of votes
+pub struct PendingOrderVotes {
+    /// Maps LedgerInfo digest to associated signatures (contained in a partial LedgerInfoWithSignatures).
+    /// This might keep multiple LedgerInfos for the current round: either due to different proposals (byzantine behavior)
+    /// or due to different NIL proposals (clients can have a different view of what block to extend).
+    li_digest_to_votes: HashMap<
+        HashValue, /* LedgerInfo digest */
```

Comment:
Now I think it is better to index by round. Each round there should be exactly one block and QC for order vote. We can avoid this hash computation.

Reply:
Indexing by round could create issues when there is a fork. I still prefer indexing by hash value. I will simplify the data structures though.